### PR TITLE
[OSD-9246] Block gaps in our AccountClaim finalizer handling

### PR DESCRIPTION
**Pull request by dkeohane**
While investigating dangling accounts I looked through our accountclaim finalize functionality to try identify areas where we may not be returning errors

Previous behaviours which have been addressed as part of this PR: 
- If the accountclaim `finalizeAccountClaim` function encountered an error, we would still proceed to remove the finalizer.
- If the `cleanUpAwsAccount` func encountered an error, we simply logged it instead of returning it.

Additions:
- Added some more unit tests. 
- Made errors from AWS visible in our logging, as debugging why we can't clean up an AWS resource without the context from AWS was impossible.

Ticket ref: [OSD-9246](https://issues.redhat.com/browse/OSD-9246)

### CODE REVIEW

**Review by mrWinston — COMMENTED** [review]
Looks good to me and the changes make sense, though i'm not  100% confident i can grasp the scope of it. Are the functions you've changed covered by the unit tests? even these small changes lead to slightly different behaviour during the reconcile, which might be what we want.

File: pkg/controller/accountclaim/reuse.go
Comment (iamkirkbater) [inline pkg/controller/accountclaim/reuse.go:45]: I think we still want to `return nil` here, as we only want to block if it's not CCS where we can't cleanup.

**Review by iamkirkbater — CHANGES_REQUESTED** [review]

File: pkg/controller/accountclaim/reuse.go
Comment (dkeohane) [reply pkg/controller/accountclaim/reuse.go:45]: Fair point - I'll update it with a comment then explaining why we're returning nil instead of the err

File: pkg/controller/accountclaim/accountclaim_controller.go
Comment (dkeohane) [inline pkg/controller/accountclaim/accountclaim_controller.go:308]: Now that we're _actually_ returning an error here, if `resetAccountSpecStatus` worked as expected but the claim wasn't cleaned up, we'll attempt to call this function again but the accountClaim.Spec.AccountLink will now be "", so we will never attempt to clean the account again and simply remove the finalizer

Comment (dkeohane) [reply pkg/controller/accountclaim/accountclaim_controller.go:308]: Actually, scratch that, resetAccountSpecStatus doesn't affect the accountclaim, just the account.

File: pkg/controller/accountclaim/accountclaim_controller.go
Comment (iamkirkbater) [inline pkg/controller/accountclaim/accountclaim_controller.go:304]: Looking at this a bit more, do we want to return err here and in the function above, or do we want to wrap err with a more descriptive error like we're doing for the isConflict error?

Comment (iamkirkbater) [reply pkg/controller/accountclaim/accountclaim_controller.go:304]: Like, we still want the original error, but we also want to know what happened inside of this error check.

File: pkg/controller/accountclaim/reuse.go
Comment (dkeohane) [inline pkg/controller/accountclaim/reuse.go:209]: As a result of moving the err var creation to outside the loop on line 192, this returned err will be the last err we retrieved from the cleangoFunc which failed

Comment (dkeohane) [inline pkg/controller/accountclaim/reuse.go:209]: As a result of moving the err var creation to outside the loop on line 192, this returned err will be the last err we retrieved from the cleangoFunc which failed

File: pkg/controller/accountclaim/reuse.go
Comment (dkeohane) [inline pkg/controller/accountclaim/reuse.go:246]: Decided to wrap these in order to gain visibility into what was coming from the AWS side of things

Comment (dkeohane) [inline pkg/controller/accountclaim/reuse.go:246]: Decided to wrap these in order to gain visibility into what was coming from the AWS side of things

File: pkg/controller/accountclaim/reuse.go
Comment (dkeohane) [reply pkg/controller/accountclaim/reuse.go:209]: Adding the return err here is possibly the most important aspect of this PR - previously if we encountered an error while cleaning up AWS resources, we would still return nil and remove the finalizer

Comment (dkeohane) [reply pkg/controller/accountclaim/reuse.go:209]: Adding the return err here is possibly the most important aspect of this PR - previously if we encountered an error while cleaning up AWS resources, we would still return nil and remove the finalizer

**Review by NautiluX — COMMENTED** [review]
/lgtm
/hold merge when ready

File: pkg/controller/accountclaim/reuse.go
Comment (iamkirkbater) [inline pkg/controller/accountclaim/reuse.go:203]: I think instead of `var err error` here we need something like `var cleanupErrors[error]{}` and append to that list, and then check if that list is failed.

Otherwise if there are multiple errors we will only see the one error message.

Functionally, this will still get what we're looking for though it might just make future debugging harder in those situations, so I won't block the PR for this request.

**Review by iamkirkbater — APPROVED** [review]
Overall LGTM.  One small comment we can address in a different PR in order to not block this one.

**Comment by iamkirkbater** [comment]
Can we squash this before we merge, please.

**Pull request merged**
